src/Parallel/Algorithm.hpp

```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
-    }
-    // set terminate to true if there are no actions in this PDAL
-    set_terminate(number_of_actions_in_phase(next_phase) == 0);
+    const PhaseType previous_phase = phase_;
     phase_ = next_phase;
-    algorithm_step_ = 0;
-    perform_algorithm();
+    bool should_start_phase = false;
+    if constexpr(detail::has_required_phase_sync_function<metavariables>) {
+      should_start_phase =
+          (not static_cast<bool>(return_phase_) or phase_ == *return_phase_ or
+           my_sync_phases_.count(phase_) != 0 or
+           metavariables::is_required_sync_phase(phase_)) and
+          number_of_actions_in_phase(phase_) != 0;
+    } else {
+      should_start_phase =
+          (not static_cast<bool>(return_phase_) or phase_ == *return_phase_ or
+           my_sync_phases_.count(phase_) != 0) and
+          number_of_actions_in_phase(phase_) != 0;
+    }
+    if (should_start_phase) {
+          if (not get_terminate() and not wait_until_next_phase_) {
+            ERROR(
+                "An algorithm must always be set to terminate at the beginning "
+                "of a phase. Since this is not the case the previous phase did "
+                "not end correctly. The integer corresponding to the previous "
+                "phase is: "
+                << static_cast<int>(previous_phase)
+                << " and the next phase is: " << static_cast<int>(next_phase));
+      }
+      if (static_cast<bool>(return_phase_) and phase_ == *return_phase_) {
+        return_phase_ = boost::none;
+      }
+      if (my_sync_phases_.count(phase_) != 0) {
+        my_sync_phases_.erase(phase_);
+      }
+      wait_until_next_phase_ = false;
+      set_terminate(false);
+      algorithm_step_ = 0;
+      perform_algorithm();
+    } else {
+      wait_until_next_phase_ = true;
+    }
   }
 
   /// Tell the Algorithm it should no longer execute the algorithm. This does
   /// not mean that the execution of the program is terminated, but only that
   /// the algorithm has terminated. An algorithm can be restarted by pass `true`
   /// as the second argument to the `receive_data` method or by calling
   /// perform_algorithm(true).
-  constexpr void set_terminate(const bool t) noexcept { terminate_ = t; }
+  constexpr void set_terminate(const bool terminate) noexcept {
+    terminate_ = terminate;
+  }
 
   /// Check if an algorithm should continue being evaluated
-  constexpr bool get_terminate() const noexcept { return terminate_; }
+  constexpr bool get_terminate() const noexcept {
+    return terminate_;
+  }
+
+  /// Tell the algorithm that on the next designated 'global sync', it should
+  /// include `phase` in the collection of phases it requests from the Main
+  /// chare.
+  ///
+  /// The global sync points are identified by calling `stop_for_sync_phases` at
+  /// a common point in the execution across elements (e.g. a slab boundary for
+  /// slab index set in the input file). During those global sync points,
```

Comment:
I thought global sync points would be given by a user-provided frequency in input file. I'm not sure what you mean when you say they're identified by calling `stop_for_sync_phases`. Can you try rephrasing this sentence?

Reply:
So, that's correct, but the way I'm planning to get that information from the input file is to have a special trigger which will ultimately call this charm entry function. I'll try to rephrase in a more user-facing form.